### ISSUE TITLE
"Pausing for 60 seconds" error when booting from USB on UEFI systems

### ISSUE THREAD

**Issue by okwizard** [puppylinux-woof-CE/woof-CE#3962]
This error is caused by the default /grub.cfg and /boot/grub.cfg files specifying pmedia=cd.
The majority of users will be booting from USB.  Solution is to specify pmedia=usbflash in these files.  To continue to support CD/DVD booting/running, a grub submenu that specifies pmedia=cd, needs to be added.  To see a working example of this you can run the recently released F96-CE_2.iso from this link:
[https://f96.puppylinux.com/](url)
I've also attached the grub menu files that were used, should be placed in /boot/grub and copy of grub.cfg put in / 
[cdmenus.tar.gz](https://github.com/puppylinux-woof-CE/woof-CE/files/10892739/cdmenus.tar.gz)

**Comment by mavrothal** [comment]
It would appear that the F96-CE_2.iso implementation is problematic when it comes to ISO booting as reported [here](https://forum.puppylinux.com/viewtopic.php?p=83604#p83604)

**Comment by dimkr** [comment]
If I understand correctly, https://github.com/puppylinux-woof-CE/woof-CE/pull/3888 should fix this by changing the meaning of `pmedia=cd` from "I want to boot from an iso9660 file system and the save file can be on a different partition" to "I want to boot from any file system and the save file can be on a different partition".

**Comment by okwizard** [comment]
@dimkr 
Would be great if the #3888 fixes it, what can I do to test it?
@mavrothal 
The OP in the forum topic you reference is booting from an unspecified OS using an undefined VM.  Neither Rockedge or myself have been able to replicate that issue in either VirtualBox or QEMU so it's hard to draw any conclusion.

Thanks
Gary

**Comment by dimkr** [comment]
> Would be great if the https://github.com/puppylinux-woof-CE/woof-CE/pull/3888 fixes it, what can I do to test it?

Unpack initrd.gz, replace /init and re-pack it.

**Comment by okwizard** [comment]
Dima, this is what was done.
1. copy code into geaney from: https://raw.githubusercontent.com/puppylinux-woof-CE/woof-CE/026bff17868a756943381218decb996cc02a800e/initrd-progs/0initrd/init
2. save as: initrd, make executable
3. unpack initrd.gz (F96-CE): 
cd to directory containing initrd.gz 
mkdir initrd-tree 
cd initrd-tree/ 
zcat ../initrd.gz | cpio -i -d 
4. replace /initrd with file from number 2
5. repack:
cd again to initrd-tree 
rm ../initrd.gz 
find . | cpio -o -H newc | gzip -9 > ../initrd.gz 
6. replace initrd.gz on USB with initrd.gz from number 5
When USB is booted it fails with Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(1.0)

Where did I go wrong?
Thanks
Gary

**Comment by CollaboratorGCM** [comment]
**@dimkr**
Would this update be in a recent vanilla?

**Comment by CollaboratorGCM** [comment]
@dimkr
Would this update be in a recent Vanilla?

**Comment by CollaboratorGCM** [comment]
@okwizard
The additional entry idea you offer seems a good one as it give users a clear sighting of what they are doing/using to boot their WoofCE distros.

Also should make it easy for developers to know when a user is reporting boot problems.

**Comment by dimkr** [comment]
> Where did I go wrong?

The file is supposed to be called "init", not "initrd".

> Would this update be in a recent vanilla?

Yes, will be released on Saturday.

**Comment by okwizard** [comment]
Well that was to obvious (smacks self in forehead), looked at it to long.  Changed and boots right up. Will report on boot test today.
Thanks
Gary

**Comment by okwizard** [comment]
@dimkr 
Dima,
I do 6 boot test using Unetbootin, Rufus and CD. To pass, the install must boot without errors, make a save file/folder, reboot and load the save file/folder.  Here's the results for your #3888 fix: 
Note that the /EFI directory was included in the ISO to allow Unetbootin to create a UEFI bootable USB.

Unet uefi = ok
Unet mbr = ok
Rufus uefi = ok
Rufus mbr = ok
CD uefi = ok
CD mbr = ok

Thanks
Gary

**Comment by okwizard** [comment]
@dimkr 
Dima,
Further testing found an issue with the USB boot mode.  Three of the four cases boot in mode 12 instead of mode 13.

Unet uefi = pupmode 12
Unet mbr = pupmode 13
Rufus uefi = pupmode 12
Rufus mbr = pupmode 12

Thanks
Gary

**Comment by mavrothal** [comment]
> Further testing found an issue with the USB boot mode. Three of the four cases boot in mode 12 instead of mode 13.

When you boot from the ISO with the vpup installation in VFAT formated disk, it finds and updates it fine but it boots to pupmode 12 instead of 13. 
This is not a major issue besides the fact that the "save" desktop icon diappears as well as the question about saving or not at the end (when activated).
Maybe when a savefile instead of a folder is used, the filesystem of the partition should also be considered.

**Comment by dimkr** [comment]
You need to boot without `pmedia=cd` or with `pmedia=usbflash` to activate PUPMODE 13. It's always been like this.

However, I'll push another commit into #3888, which switches from PUPMODE 12 to 13 when booting from a removable drive even if `pmedia=cd`, unless it's a CD. I assume most Puppy users don't use CDs, and when they do use `pmedia=cd`, they boot from a flash drive.

EDIT: 51ec5e2 seems to work nicely. I changed from `pmedia=usbflash` to `pmedia=cd` and PUPMODE 13 is (still) activated because I'm booting from a removable drive which is not a CD.

**Comment by okwizard** [comment]
@dimkr 
Dima,
Here are test for 51ec5e2.

Unet uefi = ok pm=13
Unet mbr = ok pm=13
Rufus uefi = ok pm=13
Rufus mbr = ok pm=13
CD uefi = ok pm=12
CD mbr = ok pm=12

Thanks
Gary

**Comment by dimkr** [comment]
> CD uefi = ok pm=12
> CD mbr = ok pm=12

Is that with a save file/folder on a flash drive, or some internal drive?

**Comment by okwizard** [comment]
@dimkr 
Dima,
CD uefi = ok pm=12 w/savefile on ext3 HD, 5 w/o savefile
CD mbr = ok pm=12 w/savefile on ext3 HD, 5 w/o savefile

Neither 51ec5e2 nor F96-CE_2 can load a savefile from fat32 USB

Thanks
Gary

**Comment by dimkr** [comment]
If this is an internal hard drive, then PUPMODE 12 is the expected behavior. PUPMODE 13 should be activated only when saving to a removable drive, or when booting with `pmedia=usbflash` (or something else that ends with `flash`).

**Comment by okwizard** [comment]
@dimkr
Dima,
That is exactly what is happening, the ext3 is an internal HD.  Pupmode 13 is when it boots from USB flash and the savefile is also on the USB flash drive.  What does not work is if you boot from CD and create the saveflie on a USB flash, on reboot from the CD the savefile does not load.

Thanks
Gary

**Comment by dimkr** [comment]
@okwizard Can you reproduce this issue, this time with `loglevel=7` in addition to `pmedia=cd`, then attach /initrd/tmp/bootinit.log here?

EDIT: if I understand the problem correctly, sr0 is found before the flash drive is recognized. Puppy files are found on the CD and the loop that searches for Puppy itself and the save partition stops, because it stops when Puppy files are found even if asked to search for the save partition as well:

```
while [ "$P_PART" = "" -a $NUM -le $WAITDEV ];do
```

![cond](https://user-images.githubusercontent.com/1471149/224531003-bf547915-ff7b-4a81-8d54-84ab38178630.png)

Therefore, 10e2cd2 should fix this, because it changes the condition to:

```
while [ "$LOOK_PUP" = "" -o "$P_PART" = "" ] && [ "$LOOK_SAVE" = "" -o "$SAVEPART" = "" ] && [ $NUM -le $WAITDEV ];do
```

EDIT 2: also pushed 81cc0d8. This code block is not needed if we have that loop, and if I understand it correctly, it's broken. It doesn't call get_part_info() before it calls search_func(), so the list of partitions searched for a save file is the same list used before, therefore this search is guaranteed to fail.

EDIT 3: the logic there is bad, pushed d41ad8a and 6c5bdc0

EDIT 4: simplified this PR and cleaned it up, now it's just two commits (21eeba9, 4a23874)

**Comment by okwizard** [comment]
@dimkr 
Dima, test of 6c5bdc0

The two test computers, one UEFI, one MBR both have internal HD with ext3 partition and Puppy Fossapup64 9.5 installed.

Unet uefi = #1, pm=13 w/savefile
Unet mbr = ok, pm=13 w/savefile
Rufus uefi = #1, pm=13 w/savefile
Rufus mbr = #1, pm= 13 w/savefile
CD uefi = #1, pm=12 w/savefile on ext3 internal HD 5 w/o savefile
CD mbr = #1, pm=12 w/savefile on ext3 internal HD 5 w/o savefile

CD boot can now load savefile from fat32 USB flash

#1 = if booting when no savefile exist you get a full 20 second delay

Thanks
Gary

**Comment by dimkr** [comment]
Thanks for testing and confirming that #3888 works, @okwizard! I restored the original value of `waitdev`, 5s instead of 20s.

**Comment by okwizard** [comment]
@dimkr 
Dima, test 21eeba9
Now getting "Pausing for 60 seconds" again

Thanks
Gary

**Comment by dimkr** [comment]
If you haven't applied 4a23874 on top of 21eeba9, this is expeccted

**Comment by okwizard** [comment]
@dimkr 
Dima, test of 7bba480

Unet uefi = #1, pm=13 w/savefile
Unet mbr = ok, pm=13 w/savefile
Rufus uefi = #1, pm=13 w/savefile
Rufus mbr = #1, pm= 13 w/savefile
CD uefi = #1, pm=12 w/savefile on ext3 internal HD 5 w/o savefile
CD mbr = #1, pm=12 w/savefile on ext3 internal HD 5 w/o savefile

The CD boot can now load savefile from fat32 USB flash
#1 = if booting when no savefile exist you get a full 5 second delay

This all looks good.
Thanks
Gary

**Comment by okwizard** [comment]
@dimkr 
Dima, is 7bba480 same code as 4a23874?

Thanks
Gary

**Comment by dimkr** [comment]
Yes, minus one cosmetic change.

**Comment by mavrothal** [comment]
As far as I can see vdpup10.0.157 still fails to define pupmode 13 when booting from the ISO and the savefile is on a FAT/VFAT internal drive.
Not sure if this is by design, as internal HD are not as likely to fail by the continuous read-writes, but a FAT/VFAT HD should be rather old and small and may be reasonable to minimize activity.
I think  in addition to "removable" $SAVEPART and "flash" $PMEDIA the $SAVEPART filesystem in $PUPSAVE should be considered to change PUPMODE 12 to PUPMODE 13

**Comment by dimkr** [comment]
Without my PR, PUPMODE 13 is activated only if booting with `pmedia=` and saving to a removable drive, or if `pmedia=usbflash`.

With my PR, PUPMODE 13 is activated if booting with `pmedia=` **or `pmedia=cd`** and saving to a removable drive, or if  `pmedia=usbflash`.

I think this is a slippery slope - if PUPMODE 13 should be activated when saving to FAT32, it should also be enabled for other non-journaling file systems (ext2, ext4 created with `-O ^has_journal`, etc'). This sudden change in behavior will break Puppy for many users, especially those who have very little RAM and use PUPMODE 12.

**Comment by mavrothal** [comment]
> I think this is a slippery slope - if PUPMODE 13 should be activated when saving to FAT32, it should also be enabled for other non-journaling file systems (ext2, ext4 created with -O ^has_journal, etc'). This sudden change in behavior will break Puppy for many users, especially those who have very little RAM and use PUPMODE 12.

As I said above an internal HD formatted in FAT/VFAT is a rather old beast. Especially considering you are booting from a CD. This is not necessarily the case with ext2/3/4 systems. 
In any case is not going to be a very common scenario so no strong feelings about it.

**Comment by dimkr** [comment]
This one is supposed to be fixed since 97bbc831b3706f8304da452186bd9f4a19038574. Closing, reopen if not.
